### ISSUE TITLE
Loading screen stuck at "Media"

### ISSUE THREAD

**Issue by beardeddude** [minetest/minetest#15547]
### Luanti version

```true
5.10
```

### Irrlicht device

X11

### Operating system and version

Ubuntu 22.04

### CPU model

Intel i3 10105

### GPU model

Nvidia RTX 3060

### Active renderer

OpenGL 4.6.0

### Summary

When trying to load a game the loading screen reaches "Media" and then gets stuck. The clouds continue to move in the background but I can leave it there for an hour and it will not advance. The game must be quit via multiple Ctrl-C. This issue appears to have been previously reported but was closed without a fix at https://github.com/minetest/minetest/issues/13595 . 

This issue has been around from my first install but generally it would only happen on the first load of the world. After force quitting and reopening it would work (this is currently still the behavior on my steam deck). But something changed and now I cannot get past the bug on my desktop.

I have tried:

- Deleting the .minetest folder and reinstalling game + recreating world
- Removing the app installed via the PPA and reinstalling it
- Multiple different games
- Bug is present on multiple games (Voxelibre, Minecloania)
- Bug is not present on game "Minetest world"

I tried building a debug version of luanti as suggested in the previous issue linked to get more information. I built from the stable-5 branch but was not able to load any games besides the "devtest game".

Using the debug build i get the following error but it happens immediately not once reaching "Media". So i dont think its the same bug. Am i missing something required to use mods on a debug build?

`ModError: Failed to load and run script from ../games/mineclone2/mods/ENVIRONMENT/mcl_void_damage/init.lua:
2024-12-13 15:17:41: ERROR[Main]: ...mes/mineclone2/mods/ENVIRONMENT/mcl_void_damage/init.lua:7: attempt to index global 'mcl_spawn' (a nil value)
2024-12-13 15:17:41: ERROR[Main]: stack traceback:
2024-12-13 15:17:41: ERROR[Main]:       ...mes/mineclone2/mods/ENVIRONMENT/mcl_void_damage/init.lua:7: in main chunk
2024-12-13 15:17:41: ERROR[Main]: Check debug.txt for details.`

Any help with this would be awesome. Im a dev by trade so just let me know what info you need and I will do my best to hold up my end of the effort.

### Steps to reproduce

1. Fresh install
2. Add game VoxelLibre or Mineclonia (or probably others)
3. Create new world
4. Start game
5. Loading screen stuck at "Media"

**Comment by sfan5** [comment]
> Bug is not present on game "Minetest world"

I have never heard of that. Do you mean [Minetest Game](https://github.com/minetest/minetest_game)?
Either way this suggests it's not an engine bug.

> Using the debug build i get the following error

This error suggests the game has some undeclared dependencies (see docs on `random_mod_load_order`). Nothing to do with the issue.

anyway you don't need a debug build for this, please run `luanti --verbose &>verbose.log`, reproduce the issue and share the log file here.

**Comment by beardeddude** [comment]
Thanks! That was a huge help. Im not sure how I missed the verbose flag.

So after looking at the output it was apparent it was not stuck its just very very slow. I had thought I had left it for an hour while i did something else but apparently my sense of time is terrible. The actual total load time is 15 min. This does not seem to be reduced on later runs. Every game load seems to be the same speed. 

So the real issue is "Media loads super slow resulting in 15+min load times". 

That makes sense than that simple games loaded quickly while large games have large load times.

[verbose3.log](https://github.com/user-attachments/files/18133060/verbose3.log)

**Comment by beardeddude** [comment]
Ok. I figured it out. Or at least I identified the problem code and made a hack that gets me past the bug. Its probably not a great general solution but I have no experience in C++ or this game. Hopefully you all can take this and make a general fix.

So when i started to profile the load screen (print statements, so limited detail). I found there was no lag in the Client::loadMedia method but there was lag in the ClientMediaDownloader::initialStep method. After some more profiling i found that all the lag was coming from drawLoadScreen. Basically the logic that ensures the load screen doent appear frozen is bogging it down because the draw is taking a while. I didnt look into exactly why the draw is taking so long but for some reason the draw is taking 100ms on my machine. I am able to play just fine once its loaded so i have no clue why such a simple screen is taking so long to render.

All i had to do to bypass the bug was increase the chunk_time_ms variable to an arbitrarily high number like 1000. And the entire media file set was loaded before the screen draw was called.

I get the feeling not many people experience my bug and probably the draw issue is somehow unique to my setup. Its even weirder that its seems to only be an issue on the load screen. But i would imagine that maybe a simple fix would be the following:

Just keep track of total elapsed time and if it been say 30 sec then just increase the chunk time to a high number and sacrifice frame rate to get past the load screen.

**Comment by Desour** [comment]
Can't reproduce.

You could try using a profiling tool, e.g. hotspot, see also: https://github.com/minetest/minetest/blob/master/doc/developing/misc.md

Btw, the verbose log is quite short and doesn't seem to contain a many minutes long wait time.

**Comment by beardeddude** [comment]
Here is a full log with my kinda hacky profiling added in. For some reason the draw call starts at near zero and then increases. I put a sample below. 

As stated above I havent been able to get the build with debug symbols to load mods and I need a large number of media files to reproduce the bug (error in original post). The profiler requires debug symbols.

`2024-12-15 09:51:13: VERBOSE[Main]: Client: Loaded cached media: 035dcc5710dabd9ab888a567ccabda99dc6b1961 "doc_basics_inventory.png"
2024-12-15 09:51:13: INFO[Main]: update call took 0
2024-12-15 09:51:13: INFO[Main]: checkandload took 1
2024-12-15 09:51:13: INFO[Main]: tryLoadFromCache took 1
2024-12-15 09:51:13: INFO[Main]: draw load screen took 440
2024-12-15 09:51:13: INFO[Main]: checkandload called at 23740606
2024-12-15 09:51:13: INFO[Main]: checksum call took 0
2024-12-15 09:51:13: INFO[Main]: Client::loadMedia called at 23740606
2024-12-15 09:51:13: INFO[Main]: Client::loadMedia img load time 1
2024-12-15 09:51:13: INFO[Main]: loadmedia call took 1`

[verbose.log](https://github.com/user-attachments/files/18141543/verbose.log)

**Comment by Desour** [comment]
>As stated above I havent been able to get the build with debug symbols to load mods

Ah right, I forgot about that one.
Apparently the `random_mod_load_order` setting is defaulted to true on debug builds. This is probably causing the mod error (but the game is also faulty, because it's missing a dependency, you could open a bug report there). You could explicitly disable this setting.
But for profiling you should rather build with `-DCMAKE_BUILD_TYPE=RelWithDebInfo`.

**Comment by beardeddude** [comment]
I agree sfan5. I think there is something wrong with my graphics stack (the "profiling" shows the disk loads to be fast). Thanks for the heads up on llvmpipe.I will look into that.

Considering that I doubt anyone else is going to be able to reproduce this bug and that its probably more a symptom of an underlying OS issue I think we can probably leave it here. I have a working hack to let me play.

I really appreciate all the help. There might be some value to some logic that can lower frame rate to increase load speed under the rare cases this kinda thing happens. But I understand not wanting to put in a fix for a rare bug we dont fully understand which is more a symptom than the root cause. 

This issue is now around for the rare person who might see something like this in the future.
